src/esp/gfx/CubeMap.cpp

```diff
@@ -201,64 +265,84 @@ void CubeMap::recreateTexture() {
-    }
+    optionalDepthBuffer_[cubeSideIndex] = Mn::GL::Renderbuffer{};
+    optionalDepthBuffer_[cubeSideIndex].setStorage(
+        Mn::GL::RenderbufferFormat::DepthComponent24, viewportSize);
   }
 }
 
-void CubeMap::attachFramebufferRenderbuffer() {
-  for (unsigned int index = 0; index < 6; ++index) {
-    Magnum::GL::CubeMapCoordinate cubeMapCoord =
-        convertFaceIndexToCubeMapCoordinate(index);
+void CubeMap::attachFramebufferRenderbuffer(unsigned int cubeSideIndex,
+                                            unsigned int mipLevel) {
+  Magnum::GL::CubeMapCoordinate cubeMapCoord =
+      convertFaceIndexToCubeMapCoordinate(cubeSideIndex);
 
-    if (flags_ & Flag::ColorTexture) {
-      frameBuffer_[index].attachCubeMapTexture(
-          rgbaAttachment, texture(TextureType::Color), cubeMapCoord, 0);
-    }
+  if (flags_ & Flag::ColorTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        rgbaAttachment, texture(TextureType::Color), cubeMapCoord,
+        int(mipLevel));
+  }
 
-    if (flags_ & Flag::DepthTexture) {
-      frameBuffer_[index].attachCubeMapTexture(
-          Mn::GL::Framebuffer::BufferAttachment::Depth,
-          texture(TextureType::Depth), cubeMapCoord, 0);
-    } else {
-      frameBuffer_[index].attachRenderbuffer(
-          Mn::GL::Framebuffer::BufferAttachment::Depth,
-          optionalDepthBuffer_[index]);
-    }
+  if (flags_ & Flag::VarianceShadowMapTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        vsmAttachment, texture(TextureType::VarianceShadowMap), cubeMapCoord,
+        int(mipLevel));
+  }
 
-    if (flags_ & Flag::ObjectIdTexture) {
-      frameBuffer_[index].attachCubeMapTexture(
-          objectIdAttachment, texture(TextureType::ObjectId), cubeMapCoord, 0);
-    }
+  // does NOT make any sense to talk about mip level for depth or object id
+  // texture. so the mipLevel is always 0 for both.
+  if (flags_ & Flag::DepthTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        texture(TextureType::Depth), cubeMapCoord, 0);
+  } else {
+    frameBuffer_[cubeSideIndex].attachRenderbuffer(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        optionalDepthBuffer_[cubeSideIndex]);
+  }
+
+  if (flags_ & Flag::ObjectIdTexture) {
+    frameBuffer_[cubeSideIndex].attachCubeMapTexture(
+        objectIdAttachment, texture(TextureType::ObjectId), cubeMapCoord, 0);
   }
 }
 
 void CubeMap::prepareToDraw(unsigned int cubeSideIndex,
-                            RenderCamera::Flags renderCameraFlags) {
+                            RenderCamera::Flags renderCameraFlags,
+                            unsigned int mipLevel) {
+  mipLevelSanityCheck("CubeMap::prepareToDraw():", flags_, mipLevel,
+                      mipmapLevels_);
+
+  if ((flags_ & Flag::ManuallyBuildMipmap) && (flags_ & Flag::ColorTexture)) {
+    int size = imageSize_ / pow(2, mipLevel);
```

Comment:
```suggestion
    int size = imageSize_ / (1 << mipLevel);
```

(I used to have `Math::pow2(n)` for this but ... it felt weird to have that wrapping just a bit shift, so it's no longer there.)